spec/scrigno_spec.rb

```diff
@@ -0,0 +1,412 @@
+
+  points = []
+  points << Topolys::Point3D.new(29.0, 33.1, 42.26)
+  points << Topolys::Point3D.new(29.0, 33.1, 40.80)
+  points << Topolys::Point3D.new(29.0, 36.9, 40.80)
+  points << Topolys::Point3D.new(29.0, 36.9, 42.26)
+  surfaces << {name: 's_E_wall', type: 'wall', points: points, holes: []}
+
+  points = []
+  points << Topolys::Point3D.new(24.0, 33.1, 43.00)
+  points << Topolys::Point3D.new(24.0, 33.1, 40.80)
+  points << Topolys::Point3D.new(29.0, 33.1, 40.80)
+  points << Topolys::Point3D.new(29.0, 33.1, 42.26)
+  surfaces << {name: 's_S_wall', type: 'wall', points: points, holes: []}
+
+  points = []
+  points << Topolys::Point3D.new(24.0, 33.1, 40.8)
+  points << Topolys::Point3D.new(24.0, 36.9, 40.8)
+  points << Topolys::Point3D.new(29.0, 36.9, 40.8)
+  points << Topolys::Point3D.new(29.0, 33.1, 40.8)
+  surfaces << {name: 's_floor', type: 'floor', points: points, holes: []}
+
+  # add bounding boxes
+  surfaces.each do |surface|
+    surface[:bb] = Topolys::BoundingBox.new
+    surface[:points].each {|p| surface[:bb].add_point(p)}
+  end
+
+  return surfaces
+end
+
+def create_face(surface, model)
+
+  vertices = []
+  surface[:points].each do |point|
+    vertices << model.get_vertex(point)
+  end
+  outer = model.get_wire(vertices)
+
+  holes = []
+  surface[:holes].each do |hole|
+    vertices = []
+    hole[:points].each do |point|
+      vertices << model.get_vertex(point)
+    end
+    holes << model.get_wire(vertices)
+  end
+
+  face = model.get_face(outer, holes)
+  face.attributes[:name] = surface[:name]
+  face.attributes[:type] = surface[:type]
+
+  return face
+end
+
+def default_sort(s1, s2)
+
+  type_order = {floor: 1, ceiling: 2, wall: 3, shade: 4}
+
+  if s1[:type] != s2[:type]
+    return type_order[s1[:type].to_sym] <=> type_order[s2[:type].to_sym]
+  end
+
+  if s1[:bb].minz != s2[:bb].minz
+    return s1[:bb].minz <=> s2[:bb].minz
+  end
+
+  if s1[:bb].miny != s2[:bb].miny
+    return s1[:bb].miny <=> s2[:bb].miny
+  end
+
+  if s1[:bb].minx != s2[:bb].minx
+    return s1[:bb].minx <=> s2[:bb].minx
+  end
+
+  return s1[:name] <=> s1[:name]
+end
+
+RSpec.describe Topolys do
+
+  it "can create scrigno model" do
+
+    model = Topolys::Model.new
+
+    surfaces = scrigno_surfaces
+    surfaces.sort! {|s1, s2| default_sort(s1, s2)}
+    surfaces.each do |surface|
+      create_face(surface, model)
+    end
+
+    # look for some problem vertices
+    v1 = model.find_existing_vertex(Topolys::Point3D.new(17.4,29.8,44.0))
+    v2 = model.find_existing_vertex(Topolys::Point3D.new(24,29.8,44.0))
+    v3 = model.find_existing_vertex(Topolys::Point3D.new(28,29.8,44.0))
+    v4 = model.find_existing_vertex(Topolys::Point3D.new(28.1,29.8,44.0))
+
+    expect(v1).to be_truthy
+    expect(v2).to be_truthy
+    expect(v3).to be_truthy
+    expect(v4).to be_truthy
+
+    expect(v3.id).not_to eq(v4.id)
+
+    # 1) there should be existing edges between v1<->v2 (e1) and v2<->v3 (e2), should not be one v1<->v3 (e3)
```

Comment:
@rd2 I think these three tests correspond to the output in your test, 1 and 2 are failing but I think 3 is working.